enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,325 @@
+replaces:
+superseded-by:
+---
+
+# Ingress Client TLS API
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement extends the IngressController API to allow the user to
+configure the use of client certificates for mutual TLS.
+
+## Motivation
+
+Some applications require the use of TLS with client certificates in addition to
+server certificates in order to provide two-way certificate-based authentication
+("mutual TLS", or "mTLS").  OpenShift 3 supports mTLS in the router, and the
+absence of mTLS support in OpenShift 4 blocks some users from upgrading.
+
+### Goals
+
+1. Enable the user to configure a client CA certificate bundle.
+2. Enable the user to configure optional or mandatory use of client certificates.
+3. Respect client CA certificates' certificate revocation lists.
+4. Allow the cluster administrator to restrict certificates by subject name.
+
+### Non-Goal
+
+Configuring client certificates on a per-route basis is out of scope.
+
+## Proposal
+
+The IngressController API is extended by adding an optional `ClientTLS` field
+with the eponymous type to `IngressControllerSpec`:
+
+```go
+type IngressControllerSpec struct {
+	// ...
+
+	// clientTLS specifies settings for requesting and verifying client
+	// certificates, which can be used to enable mutual TLS.
+	ClientTLS ClientTLS `json:"clientTLS,omitempty"`
+}
+```
+
+The `ClientTLS` type has a required `ClientCertificatePolicy` field of the
+eponymous type for configuring validation of client certificates, an optional
+`ClientCA` field of type `configv1.ConfigMapNameReference` for specifying a
+ConfigMap with a client CA certificate bundle, and an optional
+`AllowedSubjectPatterns` field of type `[]string` for configuring a list of
+allowed subject names:
+
+```go
+// ClientTLS specifies TLS configuration to enable client-to-server
+// authentication, which can be used for mutual TLS.
+type ClientTLS struct {
+	// clientCertificatePolicy specifies whether the ingress controller
+	// requires clients to provide certificates.  This field accepts the
+	// values "Required" or "Optional".
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	ClientCertificatePolicy ClientCertificatePolicy `json:"clientCertificatePolicy"`
+
+	// clientCA is an optional reference to a configmap containing the
+	// PEM-encoded CA certificate bundle that should be used to verify a
+	// client's certificate.  If omitted, the system trust bundle is used.
+	//
+	// +optional
+	ClientCA configv1.ConfigMapNameReference `json:"clientCA"`
+
+	// allowedSubjectPatterns specifies a list of regular expressions that
+	// should be matched against the distinguished name on a client
```

Comment:
This subject matching should be performed after the cert is verified by CA and CRL. That might not be clear from this text so we might say "on a _valid_ client certificate"